### DISCUSSION TITLE
Image display

### DISCUSSION THREAD

**Discussion by sergi0g** [Textualize/textual#1191] [Q&A] [answered]
I want to create a proper ui with textual and was wondering if (and how) it is possible to render images in the terminal. Is there a way to do that?

**Comment by davep** [comment, marked as answer]
Textual itself doesn't have direct support for anything like that at the moment, although image support [is on the roadmap](https://textual.textualize.io/roadmap/). Meanwhile though there are options such as [Rich Pixels](https://github.com/darrenburns/rich-pixels) that will turn images into `Rich` renderables. I used it it recently to [quickly show low-res versions of regions maps from Second Life in the terminal](https://www.youtube.com/watch?v=AwIPu061MHo).

**Comment by voxofox** [comment]
This is great! Thanks for the reference @davep. 

Is there perhaps interest to support raster graphics, as is made possible by e.g. the [Terminal Graphics protocol](https://sw.kovidgoyal.net/kitty/graphics-protocol/)?

For example, instead of the image rendered into the terminal columns and rows, it can render images pixel-for-pixel in a terminal which supports the protocol 
![image](https://user-images.githubusercontent.com/92116921/202885487-a3bac1e5-5bb6-41aa-b2b5-f90825a63b1d.png)

It's pretty neat!

P.s. The [pixcat](https://github.com/mirukana/pixcat) library is a good reference for how this might be integrated into a Rich / Textual module.

**Reply by davep** [reply]
> Is there perhaps interest to support raster graphics, as is made possible by e.g. the [Terminal Graphics protocol](https://sw.kovidgoyal.net/kitty/graphics-protocol/)?

I imagine that's one of the things Will had in mind when placing image support on the roadmap. :-)

**Comment by torshepherd** [comment]
It would be REALLY nice to use `notcurses`, but there are two sets of python bindings and support/documentation is really limited and slowing down. But something with that sort of rasterization speed would easily open the door to video, drawing (html canvas equivalent), and really exciting stuff. Some dude even wrote a 3d graphics library on top of notcurses...
